org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ExtendedDocumentSymbolHandler.java

```diff
@@ -0,0 +1,191 @@
+import static org.eclipse.jdt.internal.core.manipulation.JavaElementLabelsCore.F_POST_QUALIFIED;
+import static org.eclipse.jdt.internal.core.manipulation.JavaElementLabelsCore.M_POST_QUALIFIED;
+import static org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin.logInfo;
+import static org.eclipse.jdt.ls.core.internal.handlers.SymbolUtils.filter;
+import static org.eclipse.jdt.ls.core.internal.handlers.SymbolUtils.getName;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IParent;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.lsp4j.DocumentSymbol;
+import org.eclipse.lsp4j.DocumentSymbolParams;
+import org.eclipse.lsp4j.SymbolTag;
+import org.eclipse.lsp4j.jsonrpc.validation.NonNull;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.eclipse.xtext.xbase.lib.Pure;
+
+public class ExtendedDocumentSymbolHandler {
+	private PreferenceManager preferenceManager;
+
+	public ExtendedDocumentSymbolHandler(PreferenceManager preferenceManager) {
+		this.preferenceManager = preferenceManager;
+	}
+
+	public List<ExtendedDocumentSymbol> handle(DocumentSymbolParams params, IProgressMonitor monitor) {
+		ITypeRoot unit = JDTUtils.resolveTypeRoot(params.getTextDocument().getUri());
+		if (unit == null || !unit.exists()) {
+			return Collections.emptyList();
+		}
+		return getHierarchicalOutline(unit, monitor);
+	}
+
+	private List<ExtendedDocumentSymbol> getHierarchicalOutline(ITypeRoot unit, IProgressMonitor monitor) {
+		try {
+			return Stream.of(getInheritedChildren(unit, monitor)).map(child -> toDocumentSymbol(child, monitor)).filter(Objects::nonNull).collect(Collectors.toList());
+		} catch (OperationCanceledException e) {
+			logInfo("User abort while collecting the document symbols.");
+		} catch (JavaModelException e) {
+			if (!unit.exists()) {
+				JavaLanguageServerPlugin.logError("Problem getting outline for " + unit.getElementName() + ": File not found.");
+			} else {
+				JavaLanguageServerPlugin.logException("Problem getting outline for " + unit.getElementName(), e);
+			}
+		}
+		return emptyList();
+	}
+
+	private static IJavaElement[] getInheritedChildren(ITypeRoot unit, IProgressMonitor monitor) throws JavaModelException {
+		IType type = unit.findPrimaryType(); // this skips package declaration and source type
```

Comment:
I noticed that the package declaration and source type itself never gets returned because of this line. I'd like to see if I can include those elements.

This appears to be new code added to the document symbol handler at https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/2619/files#diff-820919a8fb15f7287a49d35ff449241622cb4ac6bb83580d5b214b014bea2951R162 .

Reply:
Fixed by integrating the extended document symbol logic into the regular document symbol handler logic. It's cleaner and less code.